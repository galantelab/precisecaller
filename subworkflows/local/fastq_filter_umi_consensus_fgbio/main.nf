//
// Runs FGBIO tools to filter UMI tags from FASTQ reads
//

include { FGBIO_FASTQTOBAM                  as FASTQTOBAM                  } from '../../../modules/nf-core/fgbio/fastqtobam/main'
include { FGBIO_CORRECTUMIS                 as CORRECTUMIS                 } from '../../../modules/local/fgbio/correctumis/main'
include { SAMTOOLS_BAM2FQ                   as BAM2FQ_UMI                  } from '../../../modules/nf-core/samtools/bam2fq/main'
include { BWA_MEM                           as ALIGN_UMI                   } from '../../../modules/nf-core/bwa/mem/main'
include { FGBIO_ZIPPERBAMS                  as ZIPPERBAMS_UMI              } from '../../../modules/nf-core/fgbio/zipperbams/main'
include { SAMTOOLS_BAM2FQ                   as BAM2FQ_CONS                 } from '../../../modules/nf-core/samtools/bam2fq/main'
include { BWA_MEM                           as ALIGN_CONS                  } from '../../../modules/nf-core/bwa/mem/main'
include { FGBIO_ZIPPERBAMS                  as ZIPPERBAMS_CONS             } from '../../../modules/nf-core/fgbio/zipperbams/main'
include { FGBIO_GROUPREADSBYUMI             as GROUPREADSBYUMI             } from '../../../modules/nf-core/fgbio/groupreadsbyumi/main'
include { FGBIO_CALLMOLECULARCONSENSUSREADS as CALLMOLECULARCONSENSUSREADS } from '../../../modules/nf-core/fgbio/callmolecularconsensusreads/main'
include { FGBIO_FILTERCONSENSUSREADS        as FILTERCONSENSUSREADS        } from '../../../modules/nf-core/fgbio/filterconsensusreads/main'

workflow FASTQ_FILTER_UMI_CONSENSUS_FGBIO {
    take:
    reads                     // channel: [mandatory] [ val(meta1), [ reads ] ]
    fasta                     // channel: [mandatory] [ val(meta2), [ fasta ] ]
    dict                      // channel: [mandatory] [ val(meta3), [ dict  ] ]
    bwa_index                 // channel: [mandatory] [ val(meta4), [ index ] ]
    groupreadsbyumi_strategy  // string:  [mandatory] grouping strategy - default: "Adjacency"
    min_reads                 // integer: [mandatory] the min num of reads for a consensus base/read - default: 3
    min_base_quality          // integer: [mandatory] mask (make 'N') consensus bases with Q < - default: 45
    max_base_error_rate       // real:    [mandatory] the maximum error rate for a single consensus base - default: 0.2
    umi_file                  // channel: [optional]  [ val(meta5), [ txt   ] ] - list of expected UMIs
    max_mismatches            // integer: [optional]  used with umi_file - default: 1
    min_distance              // integer: [optional]  used with umi_file - default: 2

    main:
    versions = Channel.empty()

    // Phase 1: FASTQ -> Grouped BAM

    // params.umi_read_structure is passed out as ext.args
    // FASTQ reads are converted into a tagged unmapped BAM file (uBAM)
    FASTQTOBAM(reads)

    // Our hero - or not
    ubam = FASTQTOBAM.out.bam

    // Corrects UMIs stored in BAM files when a set of fixed UMIs is in use
    if (umi_file) {
        CORRECTUMIS(ubam, umi_file, max_mismatches, min_distance)
        ubam      = CORRECTUMIS.out.bam
        versions = versions.mix(CORRECTUMIS.out.versions)
    }

    split = false // For SAMTOOLS_BAM2FQ
    sort  = false // For BWA_MEM

    // Takes an unmapped BAM and generates an aligned BAM using bwa and ZipperBams
    BAM2FQ_UMI(ubam, split)
    ALIGN_UMI(BAM2FQ_UMI.out.reads, bwa_index, [[id:'no_fasta'], []], sort)
    ZIPPERBAMS_UMI(ubam, ALIGN_UMI.out.bam, fasta, dict)

    // Group reads together that appear to have come from the same original molecule
    GROUPREADSBYUMI(ZIPPERBAMS_UMI.out.bam, groupreadsbyumi_strategy)

    // Phase 2(a): GroupedBam -> Filtered Consensus, R&D Version

    // Call consensus sequences from reads with the same unique molecular tag
    call_min_reads              = 1   // min num of reads to consensus base
    call_min_input_base_quality = 20  // filter raw reads bellow this Q
    CALLMOLECULARCONSENSUSREADS(GROUPREADSBYUMI.out.bam, call_min_reads, call_min_input_base_quality)

    // Consensus uBam -> Consensus Mapped BAM
    BAM2FQ_CONS(CALLMOLECULARCONSENSUSREADS.out.bam, split)
    ALIGN_CONS(BAM2FQ_CONS.out.reads, bwa_index, [[id:'no_fasta'], []], sort)
    ZIPPERBAMS_CONS(CALLMOLECULARCONSENSUSREADS.out.bam, ALIGN_CONS.out.bam, fasta, dict)

    // Filters consensus reads generated by CallMolecularConsensusReads
    FILTERCONSENSUSREADS(ZIPPERBAMS_CONS.out.bam, fasta, min_reads, min_base_quality, max_base_error_rate)

    versions = versions.mix(FASTQTOBAM.out.versions)
    versions = versions.mix(BAM2FQ_UMI.out.versions)
    versions = versions.mix(ALIGN_UMI.out.versions)
    versions = versions.mix(ZIPPERBAMS_UMI.out.versions)
    versions = versions.mix(GROUPREADSBYUMI.out.versions)
    versions = versions.mix(CALLMOLECULARCONSENSUSREADS.out.versions)
    versions = versions.mix(BAM2FQ_CONS.out.versions)
    versions = versions.mix(ALIGN_CONS.out.versions)
    versions = versions.mix(ZIPPERBAMS_CONS.out.versions)
    versions = versions.mix(FILTERCONSENSUSREADS.out.versions)

    emit:
    bam      = FILTERCONSENSUSREADS.out.bam     // channel: [ val(meta), [ bam ] ]
    versions = versions                         // channel: [ versions.yml ]
}

