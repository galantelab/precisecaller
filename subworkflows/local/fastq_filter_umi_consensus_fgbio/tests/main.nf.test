// nf-core subworkflows test fastq_filter_umi_consensus_fgbio
nextflow_workflow {

    name "Test Subworkflow FASTQ_FILTER_UMI_CONSENSUS_FGBIO"
    script "../main.nf"
    workflow "FASTQ_FILTER_UMI_CONSENSUS_FGBIO"

    tag "subworkflows"
    tag "subworkflows_"
    tag "subworkflows/fastq_filter_umi_consensus_fgbio"
    tag "bwa"
    tag "bwa/mem"
    tag "fgbio"
    tag "fgbio/callmolecularconsensusreads"
    tag "fgbio/correctumis"
    tag "fgbio/fastqtobam"
    tag "fgbio/filterconsensusreads"
    tag "fgbio/groupreadsbyumi"
    tag "fgbio/zipperbams"
    tag "samtools"
    tag "samtools/bam2fq"

    setup {
        run("BWA_INDEX") {
            script "../../../../modules/nf-core/bwa/index/main.nf"
            process {
                """
                input[0] = [
                    [id: 'test'],
                    file(params.local_testdata_base_path + 'references/Homo_sapiens/Gencode/hg38_chr22/Sequence/WholeGenomeFasta/genome.fa', checkIfExists: true)
                ]
                """
            }
        }
    }

    test("Homo_sapiens - fastq.gz") {

        when {
            config "./nextflow.config"

            workflow {
                """
                input[0] = [
                    [ id:'test', sample:'test_umi', read_group:'\\\'@RG\\\\tID:test\\\'' ], // meta map
                    [
                        file(params.local_testdata_base_path + 'data/Homo_sapiens/Illumina/fastq/test.umi_1.fastq.gz', checkIfExists: true),
                        file(params.local_testdata_base_path + 'data/Homo_sapiens/Illumina/fastq/test.umi_2.fastq.gz', checkIfExists: true)
                    ]
                ]
                input[1] = [
                    [ id:'genome' ],
                    file(params.local_testdata_base_path + 'references/Homo_sapiens/Gencode/hg38_chr22/Sequence/WholeGenomeFasta/genome.fa', checkIfExists: true)
                ]
                input[2] = [
                    [ id:'dict' ],
                    file(params.local_testdata_base_path + 'references/Homo_sapiens/Gencode/hg38_chr22/Sequence/WholeGenomeFasta/genome.dict', checkIfExists: true)
                ]
                input[3]  = BWA_INDEX.out.index
                input[4]  = null
                input[5]  = null
                input[6]  = null
                input[7]  = params.group_by_umi_strategy
                input[8]  = params.umi_min_reads
                input[9]  = params.umi_min_base_quality
                input[10] = params.umi_max_base_error_rate
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }
}
