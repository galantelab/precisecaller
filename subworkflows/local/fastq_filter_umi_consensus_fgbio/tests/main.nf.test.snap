{
    "Homo_sapiens - fastq.gz": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "sample": "test_umi",
                            "read_group": "'@RG\\tID:test'"
                        },
                        "test.cons.filtered.bam:md5,3b219102de4b637ae6afc7088b29d055"
                    ]
                ],
                "1": [
                    "versions.yml:md5,25a88bb0161b7c8c0776ff571ae69c14",
                    "versions.yml:md5,2eaab159bf1e0ebc230271b9b122598a",
                    "versions.yml:md5,3ac60a8fef413d5536c56531d6542d3c",
                    "versions.yml:md5,8ea9f9df1ceb8c3c85cc8758d3286677",
                    "versions.yml:md5,9bd98e354b529b1c77b81314707beaf3",
                    "versions.yml:md5,ac477a7d830567f244d6ea80f5230df4",
                    "versions.yml:md5,b09f8b4ff8f5532e891b7088734567fa",
                    "versions.yml:md5,beec23292beb0250ce1b33ba9a2e4273",
                    "versions.yml:md5,d822d6577f3cc539d920868da812cb1f",
                    "versions.yml:md5,ef394f598cdb27db7d9262f318054e7b"
                ],
                "bam": [
                    [
                        {
                            "id": "test",
                            "sample": "test_umi",
                            "read_group": "'@RG\\tID:test'"
                        },
                        "test.cons.filtered.bam:md5,3b219102de4b637ae6afc7088b29d055"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,25a88bb0161b7c8c0776ff571ae69c14",
                    "versions.yml:md5,2eaab159bf1e0ebc230271b9b122598a",
                    "versions.yml:md5,3ac60a8fef413d5536c56531d6542d3c",
                    "versions.yml:md5,8ea9f9df1ceb8c3c85cc8758d3286677",
                    "versions.yml:md5,9bd98e354b529b1c77b81314707beaf3",
                    "versions.yml:md5,ac477a7d830567f244d6ea80f5230df4",
                    "versions.yml:md5,b09f8b4ff8f5532e891b7088734567fa",
                    "versions.yml:md5,beec23292beb0250ce1b33ba9a2e4273",
                    "versions.yml:md5,d822d6577f3cc539d920868da812cb1f",
                    "versions.yml:md5,ef394f598cdb27db7d9262f318054e7b"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-09T19:04:58.232651915"
    }
}